File: 백준/Gold/13913. 숨바꼭질 4/숨바꼭질 4.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const [n, k] = input[0].split(" ").map(Number)
const MAX = 100_000
const visited = Array.from({ length: MAX + 1 }, () => -1)

const path = Array.from({ length: MAX + 1 }, () => -1)

const bfs = (start, end, visited, path) => {
  const queue = [start]
  let index = 0

  visited[start] = 0

  while (queue.length > index) {
    const cur = queue[index++]
    if (cur == end) continue

    for (let nx of [cur - 1, cur + 1, cur * 2]) {
      if (nx < 0 || nx > MAX || visited[nx] !== -1) continue
      queue.push(nx)
      visited[nx] = visited[cur] + 1
      path[nx] = cur
    }
  }
}

bfs(n, k, visited,path)
let answer = `${k}`
let start = k
for(let i=visited[k];i>0;i--){
  const next = path[start]
  answer = `${next} ` + answer
  start = next
}

console.log(visited[k])
console.log(answer)

File: 백준/Gold/13913. 숨바꼭질 4/숨바꼭질 4.py
import sys
from collections import deque
input = sys.stdin.readline

n,k = map(int,input().split())
MAX = 100_001
graph = [-1] * MAX
path = [-1] * MAX

def bfs (start,end,graph,path):
  queue = deque([start])
  graph[start] = 0

  while queue:
    cur = queue.popleft()
    if cur == end: return

    for nx in [cur-1,cur+1,cur*2]:
      if nx<0 or nx>=MAX or graph[nx] !=-1: continue
      queue.append(nx)
      graph[nx] = graph[cur] +1
      path[nx] = cur

def getPath(path,start,cnt):
  answer = [str(start)]
  index = start
  for i in range(cnt):
    nx = path[index]
    answer.append(str(nx))
    index = nx
  return ' '.join(answer[::-1])

bfs(n,k,graph,path)
print(graph[k])
print(getPath(path,k,graph[k]))

